presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddNotNullFiltersToJoinNode.java

```diff
@@ -0,0 +1,254 @@
+                inferredNotNullsVars = extractNotNullVariables(joinCriteria, joinFilter, concat(joinNode.getLeft().getOutputVariables().stream(),
+                        joinNode.getRight().getOutputVariables().stream()).collect(toList()), context.getSession());
+                break;
+            case FULL:
+                //NOT NULL cannot be inferred
+                return Result.empty();
+        }
+
+        if (inferredNotNullsVars.size() > 0) {
+            RowExpression updatedJoinFilter = buildUpdatedJoinFilter(joinFilter, buildNotNullRowExpression(inferredNotNullsVars));
+            return Result.ofPlanNode(
+                    new JoinNode(joinNode.getSourceLocation(),
+                            context.getIdAllocator().getNextId(),
+                            joinNode.getType(),
+                            joinNode.getLeft(),
+                            joinNode.getRight(),
+                            joinNode.getCriteria(),
+                            joinNode.getOutputVariables(),
+                            Optional.ofNullable(updatedJoinFilter),
+                            joinNode.getLeftHashVariable(),
+                            joinNode.getRightHashVariable(),
+                            joinNode.getDistributionType(),
+                            joinNode.getDynamicFilters()));
+        }
+
+        return Result.empty();
+    }
+
+    private boolean hasRuleAlreadyApplied(RowExpression joinFilter)
+    {
+        if (joinFilter instanceof SpecialFormExpression && ((SpecialFormExpression) joinFilter).getForm() == AND) {
+            //If the RowExpression is an AND clause
+            List<RowExpression> arguments = ((SpecialFormExpression) joinFilter).getArguments();
+            if (arguments.size() == 2) {
+                RowExpression possibleMarker = arguments.get(0);
+                //And the fist argument is our "marker", we have already applied this rule to this JoinNode
+                return NOT_NULL_MARKER_TRUE_EXPRESSION == possibleMarker;
+            }
+        }
+
+        return false;
+    }
+
+    private RowExpression buildUpdatedJoinFilter(RowExpression existingFilter, RowExpression notNullFilter)
+    {
+        //We build the updated JoinFilter with a custom "marker"
+        //We use this "marker" to determine if we have already applied this rule in #hasRuleAlreadyApplied
+        return and(NOT_NULL_MARKER_TRUE_EXPRESSION, and(existingFilter, notNullFilter));
+    }
+
+    private Collection<VariableReferenceExpression> extractNotNullVariables(List<JoinNode.EquiJoinClause> joinCriteria, RowExpression joinFilter,
+            List<VariableReferenceExpression> candidates, Session session)
+    {
+        RowExpression combinedFilter = joinFilter;
+        for (JoinNode.EquiJoinClause criteria : joinCriteria) {
+            combinedFilter = and(combinedFilter, criteria.getLeft());
+            combinedFilter = and(combinedFilter, criteria.getRight());
+        }
```

Comment:
Is there any specific reason why we need to combine all the expressions before processing?

I think the logic is getting a bit complex and can be cleaner.

If my understanding is correct, there are two sets of variables which should not be null.

- From the equijoinclause. In this case, both left and right sides are variableReferenceExpression. For inner join, we can safely add not null filter on both sides. For left join, it's right side. For right join, it's left side.
- From join filter. In this case, it's call expressions in join filters, and based on if the input of the functions can be null or not, we can add not null filter to these variables. This is a bit concerning, and we should add additional control on whether to enable it or not.

We do not need to combine them and feed to the visitor.

Reply:
To me, equijoin is just another EQUALS expression to infer NOT NULL variables from. I'm having a hard time understanding your concern. Are you up for a quick 30 min chat on {Zoom/GMeet} to build consensus ? I will reach out over email

Reply:
Sounds good. Just booked the calendar.

Reply:
Spoke offline on this. @feilong-liu's primary concern was allowing user's more fine grained control in case they wish to be more conservative on what join conditions are used for null inferencing.  
I have added a `JoinNotNullInferenceStrategy` enum with a `EQUI_JOIN_ONLY` flag to give users this control